client/src/main/java/io/pravega/client/security/auth/JwtTokenHandlingStrategy.java

```diff
@@ -0,0 +1,155 @@
+
+    /**
+     * Represents the threshold for triggering delegation token refresh.
+     */
+    @VisibleForTesting
+    static final int DEFAULT_REFRESH_THRESHOLD = 5;
+
+    private static final String REFRESH_THRESHOLD_SYSTEM_PROPERTY = "pravega.client.auth.token-refresh.threshold";
+
+    private static final String REFRESH_THRESHOLD_ENV_VARIABLE = "pravega_client_auth_token-refresh.threshold";
+
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    private final int tokenRefreshThreshold;
+
+    /**
+     * The Controller gRPC client used for interacting with the server.
+     */
+    private final Controller controllerClient;
+
+    private final String scopeName;
+
+    private final String streamName;
+
+    private final AtomicReference<DelegationToken> delegationToken = new AtomicReference<>();
+
+    JwtTokenHandlingStrategy(Controller controllerClient, String scopeName, String streamName) {
+        this(controllerClient, scopeName, streamName, ConfigurationOptionsExtractor.extractInt(
+                REFRESH_THRESHOLD_SYSTEM_PROPERTY, REFRESH_THRESHOLD_ENV_VARIABLE, DEFAULT_REFRESH_THRESHOLD));
+    }
+
+    JwtTokenHandlingStrategy(Controller controllerClient, String scopeName, String streamName, int tokenRefreshThreshold) {
+        Exceptions.checkNotNullOrEmpty(scopeName, "scopeName");
+        Preconditions.checkNotNull(controllerClient, "controllerClient is null");
+        Exceptions.checkNotNullOrEmpty(streamName, "streamName");
+
+        this.scopeName = scopeName;
+        this.streamName = streamName;
+        this.controllerClient = controllerClient;
+        this.tokenRefreshThreshold = tokenRefreshThreshold;
+    }
+
+    public JwtTokenHandlingStrategy(String token, Controller controllerClient, String scopeName, String streamName) {
+        this(token, controllerClient, scopeName, streamName, ConfigurationOptionsExtractor.extractInt(
+                "pravega.client.auth.token-refresh.threshold",
+                "pravega_client_auth_token-refresh.threshold",
+                DEFAULT_REFRESH_THRESHOLD));
+    }
+
+    public JwtTokenHandlingStrategy(String token, Controller controllerClient, String scopeName, String streamName,
+                                    int tokenRefreshThreshold) {
+        Exceptions.checkNotNullOrEmpty(token, "token");
+        Exceptions.checkNotNullOrEmpty(scopeName, "scopeName");
+        Preconditions.checkNotNull(controllerClient, "controllerClient is null");
+        Exceptions.checkNotNullOrEmpty(streamName, "streamName");
+
+        Long expTime = extractExpirationTime(token);
+        delegationToken.set(new DelegationToken(token, expTime));
+        this.scopeName = scopeName;
+        this.streamName = streamName;
+        this.controllerClient = controllerClient;
+        this.tokenRefreshThreshold = tokenRefreshThreshold;
+    }
+
+    @Override
+    public String retrieveToken() {
+        if (isTokenNearingExpiry()) {
```

Comment:
I have a general confusion/concern about the entirety of the PR, which might be based on me not understanding the overall code as a whole, or it may be my concern is legitimate. Please help me figure out which.

Here's my thing:

To my understanding, the authentication/authorization flow goes like this:
- step 1: the client initiates a conversation with the controller.   In that phase, the client calls the Credentials interface, https://github.com/pravega/pravega/blob/master/client/src/main/java/io/pravega/client/stream/impl/Credentials.java#L32 configured in the environment; depends on the plugin implementation etc.  Some decision is made there to grant an opaque token (opaque to Pravega and specific to the plugin implementation) to OK the client's intentions.
- step 2: as a result of that, a delegation token (opaque to the plugin impl. from above, and format specific to Pravega) is issued by the controller, with an expiry
- step 3: that d-token is used over and over again between the client and SS, until expiry.  

- step 4: the code above looks like essentially we would automatically refresh the d-token indefinitely, with no questions asked, no challenge.   I thought, that when the d-token expires, the client should go right back to step 1 and get challenged again to present its credentials.  (the implementation of the plugin itself can have its own meanings of caching, refreshing tokens etc etc.), oblivious of Pravega.     The d-token has its own configurable expiry, which is great and the deployer can decide what that length of time should be.  But once that runs out, shouldn't it be thrown right back to step 1?    Imagine the user initially used in step 1 no longer exists.  The d-token essentially lives on forever, always refreshing, until the end of time.  that doesn't work right?

It's possible I missed something; let me know.

Reply:
@sarlaccpit A couple of clarifications: 

1. Your description looks correct for steps 1, 2 and 3. (Assuming you were referring to the delegation token as being distinct from the token token in the Credentials interface). 
2. Regarding step 4, upon delegation token renewal the client is reauthenticated: 
https://github.com/pravega/pravega/blob/5c38d07f24f46f988c09a4345dd5d06b8bcfa770/controller/src/main/java/io/pravega/controller/server/rpc/auth/AuthInterceptor.java#L77.